### PR TITLE
FEATURE: Cleanup change projection after publish and discard

### LINKED ISSUE #4998: TASK: `ChangeProjection` should clean up orphaned change records
#### Problem description

The `ChangeProjection` in `Neos.Neos` currently misses a lot of cases in which it should clean up after itself. This leads to many orphaned `cr_*_p_neos_change` records that do not serve any purpose after the workspace they're associated with has been published or discarded.

To reproduce this problem, @bwaidelich has offered a SQL query for counting changes grouped by contentstream state via `cr_*_p_neos_change` (I'll assume `cr_default_p_neos_change` from hereon out):
```SQL
SELECT
  COUNT(*), cs.state
FROM
  cr_default_p_neos_change c
  JOIN cr_default_p_contentstream cs ON cs.contentStreamId = c.contentStreamId
GROUP BY cs.state;
```

> [!NOTE]
> Any Neos 9 instance that is in use will already have orphaned change records. To start fresh, you will need to make sure that every workspace is up-to-date and no workspace contains any changes.
> 
> Then it'll be safe to run `DELETE FROM cr_default_p_neos_change` on your database, you start at a change count of `0`. **NEVER** do this in a mission-critical environment. This is only safe in a development environment that tolerates data loss.

Now, here are the results of @bwaidelich's query for several subsequent operations (all issued by the same user):

1. Creating a document node
```
+----------+---------------------+
| COUNT(*) | state               |
+----------+---------------------+
|        2 | IN_USE_BY_WORKSPACE |
+----------+---------------------+
```
We see 2 changes, one for the document node and one for its content collection. (This may vary obviously depending on the node type)

2. Creating a content node in the content collection of that document node
```
+----------+---------------------+
| COUNT(*) | state               |
+----------+---------------------+
|        3 | IN_USE_BY_WORKSPACE |
+----------+---------------------+
```
We see 1 additional change for the creation of the content node. (Again, this may vary depending on the node type of the content node)

3. Changing a property of that content node
```
+----------+---------------------+
| COUNT(*) | state               |
+----------+---------------------+
|        3 | IN_USE_BY_WORKSPACE |
+----------+---------------------+
```
Nothing has changed. The change record for the content node has been flagged with `changed` - that's all.

4a. Discarding all changes via the UI
```
+----------+------------------+
| COUNT(*) | state            |
+----------+------------------+
|        6 | NO_LONGER_IN_USE |
+----------+------------------+
```
Here it gets interesting: The change count has doubled, but all changes are now associated with content streams that are `NO_LONGER_IN_USE`.

4b. Discarding all changes via `./flow workspace:discard [workspace-name]`
```
+----------+------------------+
| COUNT(*) | state            |
+----------+------------------+
|        3 | NO_LONGER_IN_USE |
+----------+------------------+
```
Something different happens here: Again all changes are associated with a content stream that is `NO_LONGER_IN_USE`. But the count remained at `3`. This is because the two publishing mechanism use different commands under the hood. `DiscardIndividualNodesFromWorkspace` (which happens in 4a) creates an additional intermediate contentstream, which leads to the doubling of the change count.

4c. Publishing all changes via the UI
```
+----------+------------------+
| COUNT(*) | state            |
+----------+------------------+
|        3 | FORKED           |
|        6 | NO_LONGER_IN_USE |
+----------+------------------+
```
The same phenomenon as in 4a, but there are also 3 changes in a content stream that is `FORKED`. I'm at a loss as to why that happens, but I suppose it's a detail in the handling of `PublishIndividualNodesFromWorkspace`.

4d. Publishing all changes via `./flow workspace:publish [workspace-name]`
```
+----------+------------------+
| COUNT(*) | state            |
+----------+------------------+
|        3 | NO_LONGER_IN_USE |
+----------+------------------+
```
Same as in 4b.

#### Expected behavior

- [ ] When a user publishes individual changes, those changes are deleted from `cr_*_p_neos_change`
- [ ] When a user publishes their workspace, all associated changes are deleted from `cr_*_p_neos_change`
- [ ] When a user discards individual changes, those changes are deleted from `cr_*_p_neos_change`
- [ ] When a user discards their workspace, all associated changes are deleted from `cr_*_p_neos_change`
- [ ] When a user creates a node and then deletes that node without publishing in between, the associated change is deleted from `cr_*_p_neos_change` (see also: https://github.com/neos/neos-development-collection/issues/4997)

#### Notes on solution

It will be easier to achieve the desired behavior, when the `ChangeProjection` is better able to track changes per `Workspace` rather than per `ContentSteam`.

This will require the changes proposed in https://github.com/neos/neos-development-collection/issues/4996.

### PR DESCRIPTION
**Pull request by dlubitz**
Fixes #4998

### CODE REVIEW

**Comment by kitsunet** [comment]
Very nice! This is draft because it misses handling of publish still, right?

**Comment by dlubitz** [comment]
No, because of missing tests. :D Wanted to write some more tests for the change projection.

**Review by kitsunet — APPROVED** [review]
So many tests! Thank you <3

File: Neos.Neos/Tests/Behavior/Features/Change/01-NodeCreation/01-CreateNodeAggregateWithNode_WithoutDimensions.feature [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+    When the command CreateWorkspace is executed with payload:
+      | Key                | Value            |
+      | workspaceName      | "user-workspace" |
+      | baseWorkspaceName  | "live"           |
+      | newContentStreamId | "user-cs-id"     |
+    And the command RebaseWorkspace is executed with payload:
+      | Key           | Value            |
+      | workspaceName | "user-workspace" |
+    And I am in dimension space point {}
```
Comment (mhsdesign) [inline Neos.Neos/Tests/Behavior/Features/Change/01-NodeCreation/01-CreateNodeAggregateWithNode_WithoutDimensions.feature]: This is obsolete as the workspace is fresh and we would just lose the deterministic cs id, also the dimension is set above
```suggestion
```

File: Neos.Neos/Tests/Behavior/Features/Bootstrap/ChangeProjectionTrait.php [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        }
+
+        if (count($tableRows) !== 0) {
+            $tableHeader = array_combine(array_values($table->getRow(0)), array_values($table->getRow(0)));
+            $tableRemain = $tableRows;
+            array_unshift($tableRemain, $tableHeader);
+
+            Assert::assertEmpty($tableRows, "Not all given changes where found." . PHP_EOL . (new TableNode($tableRemain))->getTableAsString());
+        }
+        Assert::assertSame(count($table->getHash()), $changes->count(), "More changes found as given.");
```
Comment (mhsdesign) [inline Neos.Neos/Tests/Behavior/Features/Bootstrap/ChangeProjectionTrait.php:73]: for debugging i always like to know the _actual_ diff like `got %s changes json_encode($changes)`

i found the `AssetUsageTrait` assert already hard to grasp :D
But ill fix this when i have problems when debugging ^^

File: Neos.Neos/Tests/Behavior/Features/Change/01-NodeCreation/02-CreateNodeAggregateWithNode_WithDimensions.feature [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+    When the command CreateWorkspace is executed with payload:
+      | Key                | Value            |
+      | workspaceName      | "user-workspace" |
+      | baseWorkspaceName  | "live"           |
+      | newContentStreamId | "user-cs-id"     |
+    And the command RebaseWorkspace is executed with payload:
+      | Key           | Value            |
+      | workspaceName | "user-workspace" |
+    And I am in dimension space point {"language": "de"}
```
Comment (mhsdesign) [inline Neos.Neos/Tests/Behavior/Features/Change/01-NodeCreation/02-CreateNodeAggregateWithNode_WithDimensions.feature]:
```suggestion
```

but we can also fix that once and for all tests (i believe asset usage has this duplicate things too?)

File: Neos.Neos/Tests/Behavior/Features/Change/07-NodeMove/02-MoveNodeAggregate_WithDimensions.feature [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+      | nodeAggregateId              | "nody-mc-nodeface"          |
+      | newParentNodeAggregateId     | "sir-nodeward-nodington-iv" |
+      | relationDistributionStrategy | "scatter"                   |
+
+    Then I expect the ChangeProjection to have the following changes in "user-cs-id":
+      | nodeAggregateId  | created | changed | moved | deleted | originDimensionSpacePoint |
+      | nody-mc-nodeface | 0       | 0       | 1     | 0       | {"language": "gsw"}       |
+    And I expect the ChangeProjection to have no changes in "cs-identifier"
+
+#  TODO: https://github.com/neos/neos-development-collection/issues/5368
```
Comment (mhsdesign) [inline Neos.Neos/Tests/Behavior/Features/Change/07-NodeMove/02-MoveNodeAggregate_WithDimensions.feature]: And these tests have to be reactivated via https://github.com/neos/neos-development-collection/issues/5368

File: Neos.Neos/Tests/Behavior/Features/Change/W01-WorkspacePublication/04-PublishIndividualNodesFromWorkspace_WithDimensions.feature [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+      | Key                             | Value                                                                                                                   |
+      | nodesToPublish                  | [{"workspaceName": "user-workspace", "dimensionSpacePoint": {"language": "de"}, "nodeAggregateId": "nody-mc-nodeface"}] |
+      | contentStreamIdForRemainingPart | "user-cs-id-remaining"                                                                                                  |
+
+    Then I expect the ChangeProjection to have the following changes in "user-cs-id-remaining":
+      | nodeAggregateId            | created | changed | moved | deleted | originDimensionSpacePoint |
+      | nody-mc-nodeface           | 1       | 1       | 0     | 0       | {"language":"gsw"}        |
+      | sir-nodeward-nodington-iv  | 1       | 1       | 0     | 0       | {"language":"de"}         |
+      | sir-nodeward-nodington-iii | 1       | 1       | 0     | 0       | {"language":"fr"}         |
+    And I expect the ChangeProjection to have no changes in "user-cs-id"
```
Comment (mhsdesign) [inline Neos.Neos/Tests/Behavior/Features/Change/W01-WorkspacePublication/04-PublishIndividualNodesFromWorkspace_WithDimensions.feature]: nice ❤️ that is the the test for the content stream was removed, splendid.

**Review by mhsdesign — COMMENTED** [review]
wow thank you sooo much for all the work ❤️ , they also work locally as expected!

And also the new `ContentStreamWasRemoved` behaviour works as quite some tests fail without ^^

id say move the tests into `PendingChange` dir, its also configured under that name

File: Neos.Neos/Tests/Behavior/Features/Change/01-NodeCreation/01-CreateNodeAggregateWithNode_WithoutDimensions.feature (hunk 1 above)
Comment (dlubitz) [reply Neos.Neos/Tests/Behavior/Features/Change/01-NodeCreation/01-CreateNodeAggregateWithNode_WithoutDimensions.feature]: Yeah, this is a left over.

File: Neos.Neos/Tests/Behavior/Features/Change/01-NodeCreation/02-CreateNodeAggregateWithNode_WithDimensions.feature (hunk 3 above)
Comment (dlubitz) [reply Neos.Neos/Tests/Behavior/Features/Change/01-NodeCreation/02-CreateNodeAggregateWithNode_WithDimensions.feature]: I removed all unnecessary rebases.

**Review by mhsdesign — APPROVED** [review]
perfect ;)

**Review by bwaidelich — COMMENTED** [review]
Awesome! Post-merge-plus-one 😀
